core/vehicle/adapter.go

```diff
@@ -0,0 +1,95 @@
+package vehicle
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/core/keys"
+	"github.com/evcc-io/evcc/server/db/settings"
+	"github.com/evcc-io/evcc/util"
+)
+
+var _ API = (*adapter)(nil)
+
+var Publish func()
+
+type adapter struct {
+	log         *util.Logger
+	name        string
+	api.Vehicle // TODO handle instance updates
+}
+
+func (v *adapter) key() string {
+	return fmt.Sprintf("vehicle.%s.", v.name)
+}
+
+func (v *adapter) publish() {
+	if Publish != nil {
+		Publish()
+	}
+}
+
+func (v *adapter) Name() string {
+	return v.name
+}
+
+// GetMinSoc returns the min soc
+func (v *adapter) GetMinSoc() int {
+	if v, err := settings.Int(v.key() + keys.MinSoc); err == nil {
+		return int(v)
+	}
+	return 0
+}
+
+// SetMinSoc sets the min soc
+func (v *adapter) SetMinSoc(soc int) {
+	v.log.DEBUG.Printf("set %s min soc: %d", v.name, soc)
+	settings.SetInt(v.key()+keys.MinSoc, int64(soc))
+	v.publish()
+}
+
+// GetLimitSoc returns the limit soc
+func (v *adapter) GetLimitSoc() int {
+	if v, err := settings.Int(v.key() + keys.LimitSoc); err == nil {
+		return int(v)
+	}
+	return 0
+}
+
+// SetLimitSoc sets the limit soc
+func (v *adapter) SetLimitSoc(soc int) {
+	v.log.DEBUG.Printf("set %s limit soc: %d", v.name, soc)
+	settings.SetInt(v.key()+keys.LimitSoc, int64(soc))
+	v.publish()
+}
+
+// GetPlanSoc returns the charge plan soc
+func (v *adapter) GetPlanSoc() (time.Time, int) {
+	var ts time.Time
+	if v, err := settings.Time(v.key() + keys.PlanTime); err == nil {
+		ts = v
+	}
+	var soc int
+	if v, err := settings.Int(v.key() + keys.PlanSoc); err == nil {
+		soc = int(v)
+	}
+	return ts, soc
+}
+
+// SetPlanSoc sets the charge plan soc
+func (v *adapter) SetPlanSoc(ts time.Time, soc int) error {
+	if !ts.IsZero() && ts.Before(time.Now()) {
+		return errors.New("timestamp is in the past")
+	}
+
+	v.log.DEBUG.Printf("set %s plan soc: %d @ %v", v.name, soc, ts.Round(time.Second).Local())
+
+	settings.SetTime(v.key()+keys.PlanTime, ts)
+	settings.SetInt(v.key()+keys.PlanSoc, int64(soc))
+
+	v.publish()
```

Comment:
Das Publishing scheint nicht zu funktionieren. Ich bekomme die Werte in der UI erst nach dem nächsten Aktualisierungsinterval. Ich verstehe aber noch nicht wieso. `publishVehicles` wird korrekt aufgerufen.

Reply:
![Nov-15-2023 14-42-53](https://github.com/evcc-io/evcc/assets/152287/3f4ef2dc-5518-4375-a65b-566a4b957d38)

Reply:
Grund ist hier, dass die Werte an der Stelle nicht aus dem Vehicle kommen sondern die `effectivePlan*` Werte vom LP kommen. Damit das klappt müssen nach setPlan nicht nur die Vehicles gepublisht werden (das klappt) sondern auch die entsprechenden Loadpoints (`lp.publishEffectiveValues()`).

@andig hast du ne gute Idee?

Reply:
Behoben